### ISSUE TITLE
(inactive) FlowSorted.Blood.WGBS.BLUEPRINT

### ISSUE THREAD

**Issue by stephaniehicks** [Bioconductor/Contributions#1207]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/stephaniehicks/FlowSorted.Blood.WGBS.BLUEPRINT

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by stephaniehicks** [comment]
This is a data package that I would like to submit to `ExperimentHub`.

**Comment by stephaniehicks** [comment]
Upon advice from @kasperdanielhansen, he suggested changing the name of the package from `FlowSorted.Blood.WGBS.BLUEPRINT` to `FlowSorted.Blood.WGBS`. I'm happy to do that, but wanted to ask best way to make that happen. Should I close this issue, change the name of the package and then open a new issue?

**Comment by mtmorgan** [comment]
@lshep can help...

**Comment by lshep** [comment]
@stephaniehicks  Change the name of the package,  update the link in the above section for repository link -  I will make the change in our database -  Please let me know when the first two are done and I'll use the updated repository link to change in the database

**Comment by stephaniehicks** [comment]
@lshep thanks for your patience! After a lengthy discussion with @kasperdanielhansen @ttriche, it has been suggested to keep the name of package as it is. I should note there was also a lengthy discussion on what file format to store the data in. Currently the object in the package loads a `BSseq` object with `loadHDF5SummarizedExperiment()`. However, @mtmorgan noted that storing the data in a simpler representation (`HDF5Array` objects) vs a derived class (e.g. `BSseq`) would allow users outside of R to use the data, which make a lot of sense. I went with the former because it takes approx 4-5 mins to create the derived class (`BSseq`) from the `HDF5Matrix` objects versus approx 15 seconds to load in the derived class.

```
 gr_complete
GRanges object with 29039352 ranges and 0 metadata columns:
             seqnames      ranges strand
                <Rle>   <IRanges>  <Rle>
         [1]     chr1 10469-10470      *
         [2]     chr1 10471-10472      *
         [3]     chr1 10484-10485      *
         [4]     chr1 10489-10490      *
         [5]     chr1 10493-10494      *
         ...      ...         ...    ...
  [29039348]     chrM 16451-16452      *
  [29039349]     chrM 16456-16457      *
  [29039350]     chrM 16497-16498      *
  [29039351]     chrM 16544-16545      *
  [29039352]     chrM 16567-16568      *
  -------
  seqinfo: 25 sequences from hg19 genome; no seqlengths
> hdf5_cov
<29039352 x 44> HDF5Matrix object of type "double":
             [,1]  [,2]  [,3] ... [,43] [,44]
       [1,]     8    20     3   .     0    28
       [2,]     6    24     3   .    13    28
       [3,]     9    18     0   .     4    22
       [4,]     8    17     2   .     4    25
       [5,]     8    20     3   .    15    23
        ...     .     .     .   .     .     .
[29039348,]  2216  2270  2497   .  1912  1332
[29039349,]  2195  2053  2463   .  1862  1233
[29039350,]  1542     0  1509   .   870   347
[29039351,]   587   251   631   .   336   132
[29039352,]    97     0   104   .    51     0
> hdf5_meth
<29039352 x 44> HDF5Matrix object of type "double":
             [,1]  [,2]  [,3] ... [,43] [,44]
       [1,]     7    15     2   .     0    13
       [2,]     5    12     3   .     9    22
       [3,]     9    13     0   .     3    13
       [4,]     8    13     2   .     4    24
       [5,]     4    14     2   .    13    20
        ...     .     .     .   .     .     .
[29039348,]     2    16    77   .    25     4
[29039349,]     0    14    71   .    24     2
[29039350,]     2     0    75   .    10     3
[29039351,]     0     1    22   .     1     2
[29039352,]     0     0    12   .     0     0
> 
> # creating in BSseq object with HDF5 matrices
> Sys.time()
[1] "2019-08-15 13:11:04 EDT"
> bs <- BSseq(gr = gr_complete, 
+             M = hdf5_meth,
+             Cov = hdf5_cov,
+             sampleNames = pheno_table$sample_name)
> Sys.time()
[1] "2019-08-15 13:16:26 EDT"
> 
> # loading in BSseq object
> Sys.time()
[1] "2019-08-15 13:16:40 EDT"
> hdf5_bs_se_path <- file.path(dataPath, "files_bsseq_hdf5_se")
> bs <- loadHDF5SummarizedExperiment(hdf5_bs_se_path)
> Sys.time()
[1] "2019-08-15 13:16:43 EDT"
```

My original thinking was someone who would work with this data in Bioconductor would likely immediately create a `BSseq` object, so it would add additional 5 mins to analysis time every time they wanted to use the data. However it was noted by @kasperdanielhansen that if I just include the simple data representations (`HDF5Array` objects), then I could make the `ExperimentHub` function such that the user who pulls the data from `ExperimentHub` would pay a 1 time cost of 5 mins to create the `BSseq` object from the `HDF5Array` objects stored on `ExperimentHub` and then caches the `BSseq` object locally. 

I would greatly appreciate your input/suggestions on if this appropriate or if I should only include simple representations. Thanks!

**Comment by Liubuntu** [comment]
Hi @mtmorgan , would you have any input on @stephaniehicks 's question? 

Also Lori @lshep will be back next week to have this data on ExpeirmentHub. 

For now, the package needs to be documented (have a NAMESPACE file and man/*.Rd files) to remove the building error.

**Comment by Liubuntu** [comment]
Hi @stephaniehicks , 

Since the users are expected to work immediately with an BSseq object, and given that the construction on the fly would take couple minutes, based on discussion with @mtmorgan (please correct me if I'm wrong here), we would suggest you have the BSseq object serialized and saved directly on the ExperimentHub. your proposal about having simple data representations (HDF5Array objects) on EH and then have a one-time construction for the BSseq object could also be an option. These will depend on EH rules and restrictions, which I am not quite sure, but @lshep could help with that. 

Best,
Qian

**Comment by Liubuntu** [comment]
## DESCRIPTION
- [ ] SUGGESTION: Add `BugReports:` field. It is encouraged to include the
  relevant links to Github for reporting Issues.
- [ ] SUGGESTION: Add `URL:` field to direct users to source code
  repositories, e.g., the GitHub repository.

**Comment by lshep** [comment]
@stephaniehicks   Let me know if anything needs to be added / removed from the hubs.

**Comment by stephaniehicks** [comment]
@Liubuntu I have done the following: 

- added a NAMESPACE file and man/*.Rd files
- added a `BugReports:` and `URL:` line to the DESCRIPTION file

@Liubuntu I understand that you suggested to uploading the serialized BSseq object to ExperimentHub. Could @mtmorgan @lshep confirm that this is the preference? I currently have the option to load it both ways (https://github.com/stephaniehicks/FlowSorted.Blood.WGBS.BLUEPRINT/blob/12c12d466e134dc1b9edcf1272420b9c43fa434a/R/FlowSorted.Blood.WGBS.BLUEPRINT.R#L45) with an argument `preloaded = TRUE` or `preloaded = FALSE`. 

@lshep -- Once I confirm which version you prefer, I will need to upload the files to ExperimentHub. Could you confirm that my credentials are the same? 

Thanks everyone! 
Stephanie

**Comment by lshep** [comment]
@stephaniehicks   Because the files are taking so long to construct, we would recommend having the serialized BSseq objects on the Hub.  Please let me know which (or if all) current hub entries should be removed and let me know when the new files (metadata.csv  and files uploaded)  are ready.  Feel free to ping me here or on slack with any hub issues.  Cheers.

**Comment by stephaniehicks** [comment]
@lshep I've updated the package to only include the serialized objects and pushed (https://github.com/stephaniehicks/FlowSorted.Blood.WGBS.BLUEPRINT/commit/79d8b9464478954a614c3ddc83240761daff9d3f). Also, I tried to upload the data using the AWS credentials that you had previously given me for earlier `ExperimentHub` package I created, but those credentials don't seem to be valid anymore. Could you email me new credentials? 

```
stephanie:FlowSorted.Blood.WGBS.BLUEPRINT shicks$ aws --profile AnnotationContributor s3 cp FlowSorted.Blood.WGBS.BLUEPRINT s3://annotation-contributor/FlowSorted.Blood.WGBS.BLUEPRINT --recursive --acl public-read
upload failed: FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/files_bsseq_hdf5_col/se.rds to s3://annotation-contributor/FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/files_bsseq_hdf5_col/se.rds An error occurred (InvalidAccessKeyId) when calling the CreateMultipartUpload operation: The AWS Access Key Id you provided does not exist in our records.
upload failed: FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/files_bsseq_hdf5_col/assays.h5 to s3://annotation-contributor/FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/files_bsseq_hdf5_col/assays.h5 An error occurred (InvalidAccessKeyId) when calling the CreateMultipartUpload operation: The AWS Access Key Id you provided does not exist in our records.
```

**Comment by stephaniehicks** [comment]
@lshep Thanks for the updated credentials! No data related to this package were previously uploaded, so no need to remove anything. I’ve just finished adding the files via AWS. Let me know when everything is in place and I’ll make sure the package is working and push any changes needed.

**Comment by lshep** [comment]
Data is in the Hub
```
> eh = ExperimentHub()
snapshotDate(): 2019-09-20
> query(eh, "FlowSorted")
ExperimentHub with 4 records
# snapshotDate(): 2019-09-20 
# $dataprovider: BLUEPRINT, Bioconductor, Bioconductor, GEO, karnanilab, GEO
# $species: Homo sapiens
# $rdataclass: RGChannelSet, character
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype 
# retrieve records with, e.g., 'object[["EH1136"]]' 

           title                                                              
  EH1136 | FlowSorted.Blood.EPIC: Illumina Human Methylation data from EPIC...
  EH2256 | FlowSorted.CordBloodCombined.450k                                  
  EH3127 | FlowSorted.Blood.WGBS.BLUEPRINT                                    
  EH3128 | FlowSorted.Blood.WGBS.BLUEPRINT (col annotation)                   
> temp = query(eh, "FlowSorted")[[4]]
> temp
An object of type 'BSseq' with
  29039352 methylation loci
  44 samples
has not been smoothed
Some assays are HDF5Array-backed
```

**Comment by stephaniehicks** [comment]
Could you remind me how to update the `snapshotDate()`? This is what I see from 

```
> library(ExperimentHub)
>  eh = ExperimentHub()
snapshotDate(): 2019-04-29
> query(eh, "FlowSorted")
ExperimentHub with 2 records
# snapshotDate(): 2019-04-29
# $dataprovider: Bioconductor, Bioconductor, GEO, karnanilab, GEO
# $species: Homo sapiens
# $rdataclass: RGChannelSet
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype
# retrieve records with, e.g., 'object[["EH1136"]]'

           title
  EH1136 | FlowSorted.Blood.EPIC: Illumina Human Methylation data from EPIC...
  EH2256 | FlowSorted.CordBloodCombined.450k
```

**Comment by lshep** [comment]
Because the package is submitted to devel - the resources will only be available when running the devel version of Bioc.

**Comment by stephaniehicks** [comment]
Ack -- of course. I'm totally not thinking today. Thanks!

**Comment by Liubuntu** [comment]
Hi @stephaniehicks , 

Your current script doesn't do the downloading (if not already cached) of EH records directly. After downloading, you may also need to reconstruct the BSseq object using the file path in the cache. The building error should be cleared when everything works correctly. 

Best,
Qian

**Comment by stephaniehicks** [comment]
Hi @Liubuntu, 

I'm not sure I understand what you are saying. The files in EH are `.h5` and `.rds` objects after serializing the `BSseq` object with the `saveHDF5SummarizedExperiment()` function. The function loads and saves using only a folder directory name. As far as I know, I cannot reconstruct the `BSseq` object with these files. Another problem is that as I noted above it will take 15 mins to reconstruct the `BSseq` object, which why it was suggested to use the serialized version. 

Could you clarify what you mean? 

Also, now that the `.h5` and `.rds` files are uploaded, it's not clear to me what to write to be able to load in these files? the `loadHDF5SummarizedExperiment()` function only accepts a directory path and does not link to the `.h5` and `.rds` files themselves? @lshep do you have suggestions? 

```
eh <- ExperimentHub()
myfiles <- query(eh, "FlowSorted.Blood.WGBS.BLUEPRINT")
myfiles[[1]]
```

```
version <- "v1.0.0"
base <- file.path("FlowSorted.Blood.WGBS.BLUEPRINT", version, "files_bsseq_hdf5_col")
loadHDF5SummarizedExperiment(base)
```

Thanks everyone!

**Comment by hpages** [comment]
Hi Stephanie,

Unfortunately `saveHDF5SummarizedExperiment()` saves an object in a form that is not convenient to host on ExperimentHub.

Here is why:

Saving a BSseq object with `saveHDF5SummarizedExperiment()` generates a folder with 2 files in it: `assays.h5` and `se.rds`. An important thing to keep in mind is that `se.rds` contains the original serialized BSseq object but without the assay data in it. The assay data is in `assays.h5`.

It seems that you've uploaded these 2 files to ExperimentHub (resources EH3127, and EH3128). I can get these resources with:
```
library(ExperimentHub)
eh <- ExperimentHub()
path_to_assays_h5 <- eh[["EH3127"]]
bs <- eh[["EH3128"]]
```

`path_to_assays_h5` is the path to a standalone 2.6G HDF5 file that contains the datasets of the 2 assays:
```
> h5ls(path_to_assays_h5)
  group     name       otype dclass           dim
0     / assay001 H5I_DATASET  FLOAT 29039352 x 44
1     / assay002 H5I_DATASET  FLOAT 29039352 x 44
```

The 1st (minor) issue here is that the names of the datasets (`assay001` and `assay002`) are automatic names that were created by `saveHDF5SummarizedExperiment()`. This is not really an issue when `loadHDF5SummarizedExperiment()` is used to reconstruct the original BSseq object because the end-user will typically operate on this BSseq object at a level where the exact names of the HDF5 datasets won't be visible. However these names are not great in the context of someone working directly with the HDF5 file. Something like `meth` and `coverage` would be better names. In other words, the HDF5 file is a valuable standalone resource per-se and should be as self-explanatory as possible.

Another (more serious) issue is that `bs` is a broken object:
```
> assay(bs)
<29039352 x 44> DelayedMatrix object of type "double":
HDF5-DIAG: Error detected in HDF5 (1.10.5) thread 0:
  #000: H5F.c line 509 in H5Fopen(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #001: H5Fint.c line 1498 in H5F_open(): unable to open file: time = Wed Sep 25 00:19:20 2019
, name = 'assays.h5', tent_flags = 0
    major: File accessibilty
    minor: Unable to open file
  #002: H5FD.c line 734 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #003: H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = 'assays.h5', errno = 2, error message = 'No such file or directory', flags = 0, o_flags = 0
    major: File accessibilty
    minor: Unable to open file
Error in h5mread(filepath, name, starts = index) : 
  failed to open file 'assays.h5'
```

That's because the object has been separated from its `assays.h5` companion.

One thing to keep in mind is that an object saved with `saveHDF5SummarizedExperiment()` needs to be loaded back into R with `loadHDF5SummarizedExperiment()`. But the `loadHDF5SummarizedExperiment()` function itself can only be pointed to a folder that is organized in the way that `saveHDF5SummarizedExperiment()` organized it, that is, with the `assays.h5` and `se.rds` files in it.

Even though it would be possible for your `FlowSorted.Blood.WGBS.BLUEPRINT()` function to (1) download the 2 files from ExperimentHub (to the local ExperimentHub cache), (2) create a temporary directory, (3) copy and rename the 2 files from the local ExperimentHub cache to the temporary directory, and (4) finally point `loadHDF5SummarizedExperiment()` to this temporary directory, this solution would be inefficient and fragile.

A better approach is to upload to ExperimentHub whatever components need to be passed to the `BSseq()` constructor to create the object, that is:
  - The HDF5 file (already on ExperimentHub but maybe you want to consider changing the dataset names).
  - The rowRanges i.e. the GRanges object passed to the `gr` argument of `BSseq()`.
  - The sample names: this could be a serialized character vector but it would make a lot of sense to store it in the same HDF5 file as the assay data (as a 3rd dataset).

Then `FlowSorted.Blood.WGBS.BLUEPRINT()` can simply be something like:
```
FlowSorted.Blood.WGBS.BLUEPRINT <- function()
{
    eh <- ExperimentHub()
    assays_h5file <- eh[["some_EH_ID"]]
    gr <- eh[["another_EH_ID"]] 
    M <- HDF5Array(assays_h5file, "M")
    Cov <- HDF5Array(assays_h5file, "Cov")
    sampleNames <- as.character(HDF5Array(assays_h5file, "sample_names"))
    BSseq(M, Cov, gr=gr, sampleNames=sampleNames)
}
```

You mentioned earlier that this is very slow and indeed it is. This is because the `BSseq()` constructor function validates the assays i.e. it checks that `all(0 <= M <= Cov) && !anyNA(M) && !anyNA(Cov) && all(is.finite(Cov))` (this check is implemented in C++ in `bsseq/src/check_M_and_Cov.cpp`). This means that all the data in the HDF5 file is read and checked, which of course takes a long time. For curated/trusted datasets like yours, it's fair to assume that the data has been checked before being uploaded to ExperimentHub so validating it again every time a user calls `FlowSorted.Blood.WGBS.BLUEPRINT()` seems unnecessary.

We should ask @kasperdanielhansen or @PeteHaitch if the bsseq package provides a way to construct a BSseq object from trusted assays i.e. without validating them. If not, maybe this could be a reasonable request. E.g. this could be supported by adding a `check` argument to the `BSseq` and `bsseq:::.BSseq` constructors and calling `new2("BSseq", ..., check=check)` instead of `new("BSseq", ...)` in the latter.

In the meantime, a workaround is to replace the call to `BSseq()` with:
```
FlowSorted.Blood.WGBS.BLUEPRINT <- function()
{
    ...
    se <- SummarizedExperiment(list(M=M, Cov=Cov),
                               rowRanges=gr,
                               colData=DataFrame(row.names=sampleNames))
    new2("BSseq", se, check=FALSE)
}
```

The FlowSorted.Blood.WGBS.BLUEPRINT package should import the ExperimentHub package and depend on S4Vectors, HDF5Array, SummarizedExperiment, and BSseq. The NAMESPACE file should import these 5 packages and export the `FlowSorted.Blood.WGBS.BLUEPRINT` function.

Hope this helps,
H.

**Comment by PeteHaitch** [comment]
@hpages The internal construction `bsseq:::.BSseq()` will skip validation.

**Comment by hpages** [comment]
Thx Pete. Are you sure `bsseq:::.BSseq()` skips validation? It uses `new()`, not `new2()`.

**Comment by PeteHaitch** [comment]
Ugh you're right, @hpages! I misremembered. Sorry for the confusion. I'll think about the best way of adding a constructor that skips validity checking. Any pointers to existing constructors that offer this option? And I'll also think about just removing this validity check entirely.

**Comment by stephaniehicks** [comment]
@hpages -- Thank you so much for your helpful comments! This all makes sense. I've have implemented the changes you described. Specifically: 

- Now using the `M` and `Cov` HDF5 files, the GRanges object, and a colData object with phenotypic information. 
- I'm using the work around you suggest to avoid the costly step with the validation. @PeteHaitch I'm happy to switch over to `bsseq:::.BSseq()` once you tell me that's what I should use. For now I'm using @hpages's trick with `new2()` (see above). 
- The 5 packages and the function name are included in the `NAMESPACE`
- @lshep I've uploaded new files via AWS. You can delete the files that were previously there (specifically the folder `files_bsseq_hdf5_col` in the path `FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/files_bsseq_hdf5_col` which contained the serialized objects. I've now got three files in `FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/`. One thing I am confused about it was as I was uploading the files, it said that I was uploading 19.2 GB of data:

```
[compute-068 /users/shicks1/myRpkgs/tmp_data]$ aws --profile AnnotationContributor s3 cp FlowSorted.Blood.WGBS.BLUEPRINT s3://annotation-contributor/FlowSorted.Blood.WGBS.BLUEPRINT --recursive --acl public-read
upload: FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/blueprint_blood_colData.RDS to s3://annotation-contributor/FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/blueprint_blood_colData.RDS
upload: FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/blueprint_blood_gr.RDS to s3://annotation-contributor/FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0/blueprint_blood_gr.RDS
Completed 2.9 GiB/19.2 GiB (81.6 MiB/s) with 1 file(s) remaining   
```

Could you confirm that you are seeing 19.2 GB of data on `ExperimentHub`? When I look at the size of the data stored locally, I see it's more like 7.3 GB. Am I uploading correctly?

```
[compute-068 /users/shicks1/myRpkgs/tmp_data]$ du -sh *
7.3G	FlowSorted.Blood.WGBS.BLUEPRINT
```

or broken up by the three files: 
```
[compute-068 /users/shicks1/myRpkgs/tmp_data/FlowSorted.Blood.WGBS.BLUEPRINT/v1.0.0]$ du -sh *
19K	blueprint_blood_colData.RDS
71M	blueprint_blood_gr.RDS
7.2G	blueprint_blood.h5
```

Looking forward to your feedback. Thank you again!

**Comment by hpages** [comment]
@stephaniehicks I'm surprised that your new h5 file is 7.2G given that the former one was 2.6G. Are you sure the new one is chunked and compressed like the former one was? Also I'm not sure what's in your `blueprint_blood_colData.RDS` file but, if this is just a serialized character vector containing the sample names, then it would be easy to store it in the h5 file along with the 2 other datasets (see my earlier comment about this). This would be slightly better than having it as a separate EH resource.

@PeteHaitch We have many low-level constructors in the S4Vectors/IRanges/GenomicRanges stack that skip validation. None of them is exported though (they're intended for internal use only). See for example the low-level DFrame constructor in S4Vectors (`S4Vectors:::new_DataFrame`, I should probably rename it `S4Vectors:::new_DFrame`). I would say that the BSseq situation is kind of unique though because AFAIK it's the only SummarizedExperiment derivative with a validity method that validates the _content_ of the assays (the validity method for SummarizedExperiment objects only validate their shape).

**Comment by hpages** [comment]
@stephaniehicks Another way to reduce the size of your h5 file is to store the data as INTEGER instead of FLOAT. You can do this by making sure that the _storage mode_ of your matrices is `"integer"` instead of `"double"` before you write them to the HDF5 file (if it's not, set it with `storage.mode(m) <- "integer"`). This will also make block processing slightly more efficient later on when you operate on the HDF5Array object because the matrix blocks will also use integers instead of doubles so their memory footprint will be reduced by half (4 bytes per integer vs 8 bytes per double).

**Comment by Liubuntu** [comment]
Thanks @hpages for the clear explanation. That makes a lot more sense to me also. Sorry @stephaniehicks that I didn't give more useful suggestions by only identifying the problem. Obviously  I have learned that the `loadHDF5SummarizedExperiment` could not reload a serialized object from EH directly. I have actually spent quite some time yesterday trying to figure out a right way to do it but failed due to unfamiliarity of both `BSseq` and `HDF5SummarizedExperiment`.  

Thanks everyone!

**Comment by hpages** [comment]
@Liubuntu No problem. Happy to help!

@stephaniehicks After giving it one more round of thoughts, I would make the case for also storing the genomic ranges in the h5 file. That way the h5 file becomes a standalone, self-contained and self-describing resource that you can share directly with your co-workers, even those who don't use ExperimentHub or R. Otherwise `blueprint_blood_gr.RDS` will be an EH resource hanging in the air with no proper meaning unless it's used together with the other EH resource that is the h5 file.

An easy way to add it to the h5 file is to split it in 2 datasets, one for the chromosome names (e.g. `row_chroms`) and one for the ranges (e.g. `row_ranges`). The first one can be of type STRING (like the `sample_names` dataset) and the second one can be an Nx2 INTEGER matrix where N is the number of rows in the assays (the 1st and 2nd columns are the 1-based starts and ends of the ranges, respectively). Hope this makes sense.

If this BS-seq dataset is the 1st of a long series of BS-seq datasets that are going to end up on EH, it's worth spending some time trying to figure out how to do this the right way. Also by establishing some convention on how to organize the data on EH, you'll be able to re-use more code instead of having to come up with ad-hoc code for each dataset. For example, you'll be able to replace `FlowSorted.Blood.WGBS.BLUEPRINT()` with a more generic helper function for making a BS-seq dataset from an h5 file e.g.:
```
makeBSseqFromHDF5File <- function(filepath)
{
    M <- HDF5Array(filepath, "M")
    Cov <- HDF5Array(filepath, "Cov")
    row_chroms <- as.character(HDF5Array(filepath, "row_chroms"))
    row_ranges <- as.matrix(HDF5Array(filepath, "row_ranges"))
    gr <- GRanges(row_chroms, IRanges(start=row_ranges[ , 1L], end=row_ranges[ , 2L]))
    sampleNames <- as.character(HDF5Array(filepath, "sample_names"))
    se <- SummarizedExperiment(list(M=M, Cov=Cov),
                               rowRanges=gr,
                               colData=DataFrame(row.names=sampleNames))
    new2("BSseq", se, check=FALSE)
}
```
Note that people that got your h5 file via a different mean than EH can use `makeBSseqFromHDF5File()` to read it.

Then `FlowSorted.Blood.WGBS.BLUEPRINT()` simply becomes a thin wrapper around `makeBSseqFromHDF5File()`:
```
FlowSorted.Blood.WGBS.BLUEPRINT <- function()
{
    eh <- ExperimentHub()
    filepath <- eh[["some_EH_ID"]]
    makeBSseqFromHDF5File(filepath)
}
```

Not sure where `makeBSseqFromHDF5File()` should live though to make it easy to re-use. In the bsseq package? Would be good to get some input from @kasperdanielhansen and @PeteHaitch about this proposal (`makeBSseqFromHDF5File()` function and proposed format for h5-based BS-seq datasets).

Thanks!

**Comment by lshep** [comment]
Sorry I missed this and wanted to check in -  @stephaniehicks   do you plan on making any other changes to the three data files after @hpages  comments or should I move forward with putting these into the hub?    Also verifying the two previous entries should be removed from the hub?

**Comment by stephaniehicks** [comment]
Hi @lshep @hpages @Liubuntu -- I apologize for dropping the ball on this. This week was much busier than I anticipated. I will get this turned around by Monday.

**Comment by Liubuntu** [comment]
Hi @stephaniehicks , 

Just a reminder that the [release schedule](http://bioconductor.org/developers/release-schedule/) indicates that the last day to accept new packages into Bioc3.10 would be next Wednesday, so that we can have this data available in the new release. Thanks!

Qian 

```
Wednesday October 23

Deadline to add new packages to the BiocC 3.10 manifest. Package submitted to tracker must have
completed the review processes and been accepted to be added to the manifest
```

**Comment by stephaniehicks** [comment]
@Liubuntu -- Thank you for the kind reminder. At the moment, there seems to be problem with working with HDF5 files on my cluster. It's taken me a week to figure out that what I was seeing was not a problem with any Bioconductor packages, but rather my cluster itself. I'm assuming the problem with my cluster will not be sorted by Wednesday, which is unfortunate. I will continue to working on this though in the mean time and will definitely have this sorted out soon, but it won't likely make the 3.10 release. I will have to wait until the 3.11 release. I'm assuming this is still OK?

**Comment by Liubuntu** [comment]
@stephaniehicks Yes, it's OK. If it couldn't make for this release, it will go to the 3.11 devel between the curernt release and next April/May until it's 3.11 release.

**Comment by lshep** [comment]
What's the status of this package?  Should we close it for now?

**Comment by stephaniehicks** [comment]
Thanks for the message @lshep. Feel free to close it for now. I was a bit back logged on other projects needed to be finished. 

I do plan to getting this package submitted by this month and I'm happy to open a new issue (or re-open this issue -- whatever is preferred) at that time. Just let me know.

**Comment by lshep** [comment]
Please request to have this issue reopened here.   Thanks
